core/interfaces/i_draggable.ts

```diff
@@ -4,11 +4,63 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-// Former goog.module ID: Blockly.IDraggable
+import {Coordinate} from '../utils/coordinate';
+import {IDragTarget} from './i_drag_target';
 
-import type {IDeletable} from './i_deletable.js';
+/**
+ * Represents an object that can be dragged.
+ */
+export interface IDraggable extends IDragStrategy {
+  /** Returns true iff the element is currently movable. */
+  isMovable(): boolean;
+
+  /** Returns the current location of the draggable in workspace coordinates. */
+  getLocation(): Coordinate;
+}
 
 /**
- * The interface for an object that can be dragged.
+ * Represents an object that can be dragged.
  */
-export interface IDraggable extends IDeletable {}
+export interface IDragStrategy {
+  /**
+   * Handles any drag startup (e.g moving elements to the front of the
+   * workspace).
+   *
+   * @param e PointerEvent that started the drag; could be used to
+   *     check modifier keys, etc.  May be missing when dragging is
+   *     triggered programatically rather than by user.
+   */
+  startDrag(e?: PointerEvent): void;
+
+  /**
+   * Handles moving elements to the new location, and updating any
+   * visuals based on that (e.g connection previews for blocks).
+   *
+   * @param newLoc Workspace coordinate to which the draggable has
+   *     been dragged.
+   * @param e PointerEvent that continued the drag; could be used to
+   *     look up any IDragTarget the pointer is over, check modifier
+   *     keys, etc.
+   * @param target The drag target the pointer is over, if any.  Could
+   *     be supplied as an alternative to providing a PointerEvent for
+   *     programatic drags.
+   */
+  drag(newLoc: Coordinate, e?: PointerEvent): void;
```

Comment:
If you want to structure the method signature like this, should the event be non-optional? Otherwise you basically have three function signatures:

```
drag(newLoc: Coordinate);
drag(newLoc: Coordinate, e: PointerEvent);
drag(newLoc: Coordinate, target: IDragTarget);
```

Reply:
That looks like what I intended.

If `e: PointerEvent` is non-optional then `target` would need to be nullable (or otherwise optional) to cover the case where you are doing a programmatic drag (no `PointerEvent`, unless maybe you synthesise one) and you don't happen to be over a drag target.